Реализация базы данных приложения

sarychv opened the pull request:
В файле README.md приведены ER диаграмма и ее описание для базы данных приложения, описание полей таблиц, несколько SQL запросов для получения информации о фильмах и пользователях сервиса.

sarychv committed:
Описание схемы базы данных приложения

В файле README.md приведены ER диаграмма и ее описание для базы данных
приложения, описание полей таблиц, несколько SQL запросов для получения
информации о фильмах и пользователях сервиса.

sarychv committed:
fix: Описание схемы базы данных приложения: добавлена сущность rating

В файле README.md приведены ER диаграмма и ее описание для базы данных
приложения, описание полей таблиц, несколько SQL запросов для получения
информации о фильмах и пользователях сервиса.

IHukumka commented:
хорошо отрисована схема, взаимосвязи понятны и не запутаны.
описание принципов работы информативно, но не хватает полной структуры взаимодействия Приложения с БД, т.е. процесса десериализации объекта из БД.

IHukumka commented:
реализация друзей через две таблицы кажется громоздкой (на мой взгляд), для получения статуса заявки необходимо проверить две таблицы (во избежание потери данных).

sarychv committed:
fix: изменено представление связи Дружба (friendship)

Исключена таблица make_friend, таблица friends преобразована в friendship,
в которую также введен дополнительный столбец is_confirmed.

sarychv committed:
TZ11

sarychv committed:
TZ11. fix: add whitespace

sarychv committed:
TZ11. fix: defined mpa field in FilmTest class

sarychv committed:
TZ11. fix: change initial block to constructor

sarychv committed:
TZ11. fix: insert new line

sarychv committed:
TZ11. fix: пытаюсь устранить какую-то бредовую ошибку при проверке синтаксиса

ErkinKasymaliev commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java:
```diff
@@ -0,0 +1,210 @@
+    }
+
+    @Override
+    public void add(Film film) {
```
Этот метод лучше сделайте private Film add(), чтобы возвращал объект с id.

SarychV replied:
Эркин, привет! 
Да, то, что ты предлагаешь было бы лучше, если бы я сразу реализовал возврат объетов классов User и Film в методах add() и update()  в интерфейсах FilmStorage() и UserStorage(). Необходимость возврата этих объектов я понял уже в конце выполнения задания. Когда вместо возврата поступивших в контроллеры объектов (как это предлагается в теории курса) я решил возвращать объекты из базы (мне так видится более правильным). Однако, реализация того, что ты предлагаешь в этом и последующих замечаниях потребует внесения изменений во все тесты, в классы с работой с базой данных и в классы по работе с хранилищем в памяти. Это достаточно большой объем работы. С учетом жесткого дедлайна, я не решился на внесение этих изменений, поскольку работа практически была оттестирована. Я предлагаю для данной реализации оставить ситуацию такой, какой она есть. Если ты считаешь вопрос принципиальным, то я предлагаю эти изменения сделать в следующеий итерации, сейчас у меня нет времени (с учетом того, что мне поставили срок сдать работу до конца завтрашнего дня), мне могут предложить вязть перенос обучения, если я не успею закрыть эту работу. К тому же, мне видится  этот вопрос не принципиальным, хотя еще раз повторю, что в целом я с тобой согласен.

ErkinKasymaliev replied:
привет, да без проблем, этот момент оставляй как есть

ErkinKasymaliev commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -30,17 +30,21 @@ public FilmController(FilmService service) {
     @PostMapping
     public Film addFilm(@Valid @RequestBody Film film) {
+        log.info("В фильмотеку добавляется фильм: " + film);
         filmService.addFilm(film);
```
метод filmService.addFilm(film) сделайте не void, а чтобы возвращал объект с генерированным id. И вам не нужен будет лишний вызов в БД на 35 строке. Должно быть примерно так:
```
public Film addFilm(@Valid @RequestBody Film film) {
        log.info("В фильмотеку добавляется фильм: " + film);
        Film savedFilm  = filmService.addFilm(film);
        log.info("Добавлен фильм: " + savedFilm);
        return savedFilm;
}
```
аналогично для addUser в UserController

SarychV replied:
см. первый комментарий.

ErkinKasymaliev commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -30,17 +30,21 @@ public FilmController(FilmService service) {
         filmService.updateFilm(film);
-        log.info("Фильм обновлен в фильмотеке: " + film);
-        return film;
+        Film modifiedFilm = filmService.findFilmById(film.getId());
```
здесь тоже лишний вызов, попробуйте так:
```
...{
   log.info("Вносятся изменения в фильм:" + film);
   Film modifiedFilm = filmService.updateFilm(film);
   log.info("Фильм изменен: " + modifiedFilm);
   return modifiedFilm;
}
```
аналогично для updateUser в UserController

SarychV replied:
см. первый комментарий.

ErkinKasymaliev commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -30,17 +30,21 @@ public FilmController(FilmService service) {
 
     @PutMapping
     public Film updateFilm(@Valid @RequestBody Film film) {
         hasValidFilmId(film.getId());
```
такие проверки лучше убрать в сервис

SarychV replied:
Исполнил.

ErkinKasymaliev commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java:
```diff
@@ -0,0 +1,210 @@
+    }
+
+    @Override
+    public void update(Film film) {
```
лучше этот метод лучше сделайте private Film update(), чтобы возвращал объект с id.

SarychV replied:
см. первый комментарий.

ErkinKasymaliev commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java:
```diff
@@ -0,0 +1,210 @@
+        List<Film> films = new ArrayList<>();
+        String sqlQuery = "SELECT id FROM films ORDER BY id ASC;";
+        List<Integer> ids = jdbcTemplate.query(sqlQuery, (rs, num) -> rs.getInt("id"));
+        for (int id : ids) {
```
Представьте что у вас 100 000 фильмов в бд, и здесь у вас будет еще 100 000 запросов по id. Это нехорошая практика, вместо этого в sqlQuery добавьте join чтобы в одном запросе подтянуть рейтинги и жанры

SarychV replied:
Еще нужно подтянуть лайки. Посмотри, пожалуйста, я реализовал то что требуется, или нет?

ErkinKasymaliev commented on src/main/java/ru/yandex/practicum/filmorate/storage/UserDbStorage.java:
```diff
@@ -0,0 +1,155 @@
+    public Collection<User> findAllUsers() {
+        List<User> users = new ArrayList<>();
+        String sqlQuery = "SELECT id FROM users;";
+        List<Integer> ids = jdbcTemplate.query(sqlQuery, (rs, num) -> rs.getInt("id"));
```
аналогичный коммент как и в FilmDbStorage. добавьте join-ы в запрос

ErkinKasymaliev reviewed (changes requested):
привет, хорошая работа!
Правда есть моменты где надо улучшить и оптимизировать код

sarychv committed:
TZ11. After review#1

ErkinKasymaliev commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java:
```diff
@@ -0,0 +1,305 @@
+    @Override
+    public Collection<Film> findAllFilms() {
+        String sqlQuery =
+        "SELECT id, name, description, release_date, duration, rating_id, rating_name, genre_id, genre_name, user_id " +
```
хим, думаю можно так:
```
select f.*, r.name, g.name from films f
join rating r on r.film_id=f.id
join film_genre fm on fm.film_id = f.id
join genre g on g.id=fg.genre_id
```
и закоменченный код удали

ErkinKasymaliev replied:
вместо f.* можешь нужные поля вставить, а не все подтягивать. И думаю, что лайки тут не нужны

ErkinKasymaliev commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -2,21 +2,21 @@
 
     @Autowired
-    public FilmService(InMemoryFilmStorage storage) {
+    public FilmService(FilmDbStorage storage) {
```
вот тут проглядел. Здесь используй @Qualifier.
вот так:
public FilmService(@Qualifier(name="filmDbStorage") FilmStorate storage) - надо инжектить абстракцию, то есть интерфейс. Прочитай про SOLID, как будет время

ErkinKasymaliev commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -14,11 +14,10 @@
 
     @Autowired
-    public UserService(InMemoryUserStorage storage) {
+    public UserService(UserDbStorage storage) {
```
здесь также
вот так:
public UserService(@Qualifier(name="userDbStorage") UserStorage storage)

ErkinKasymaliev commented on src/main/java/ru/yandex/practicum/filmorate/service/MpaGenreService.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.Collection;
+
+@Service
+public class MpaGenreService {
```
не надо так делать: раздели на отдельные сервисы MpaService, GenreService. также со storage-ми

ErkinKasymaliev reviewed (approved):
привет, работа принята!
Оставил дополнительные комменты. Если будешь успевать, то поправь их перед мержем. Если же нет - то к следующей домашку
